### ISSUE TITLE
F811 false positive in match statement

### ISSUE THREAD

**Issue by cgahr** [charliermarsh/ruff#4076]
Hi,

really love ruff, such a great tool! It also inspired me to learn Rust ;)

I found the following bug, which can be reproduced with this MWE:
```python
def no_redef(value):
    if value:

        def fun(x, y):
            return x

    else:

        def fun(x, y):
            return y

    return fun


def redef(value):
    match value:
        case True:

            def fun(x, y):
                return x

        case False:

            def fun(x, y):
                return y

    return fun
```
Running ruff throws the following error which (IMO) is a bug:
```bash
> ruff redef.py --isolated
redef.py:24:17: F811 Redefinition of unused `fun` from line 19
Found 1 error.
```
**Why is this a bug?** 
In my understanding both the `no_redef` and `redef` function have the same behavior. Thus, F811 should behave the same for both functions, which is not the case.

This behavior is consistent with flake8/pyflakes, however, I think this is a bug in their code as well and I opened a ticket there too (https://github.com/PyCQA/pyflakes/issues/771).

**Ruff version**
```bash
> ruff --version
ruff 0.0.261
```

**pyproject.toml**
```toml
[tool.ruff]
select = ["F"]
target-version = "py38"
```

Constantin

**Comment by JonathanPlasse** [comment]
I think the solution would be to add `MatchCase` to the binding and implement an equivalent `different_forks` for `MatchCase` instead of `Stmt`.
I can pick it up.
